### PR TITLE
Add leapfrog support to MNAIO

### PR DESCRIPTION
**Pull request by git-harry**
https://rpc-openstack.atlassian.net/browse/LA-322
https://rpc-openstack.atlassian.net/browse/LA-324

Issue: [LA-322](https://rpc-openstack.atlassian.net/browse/LA-322)

### CODE REVIEW

File: rpc_jobs/multi_node_aio.yml [hunk 1]
```diff
@@ -1,37 +1,145 @@
...
+    image:
       - xenial:
           DEFAULT_IMAGE: "16.04.2"
       - trusty:
           DEFAULT_IMAGE: "14.04.5"
+    action:
+      - deploy:
+          ACTION_STAGES: >-
+            Prepare MaaS,
+            Setup MaaS,
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: You probably want to remove the maas stages, and use DEPLOY_MAAS: yes as default for jobs in the job template.

File: rpc_jobs/multi_node_aio.yml [hunk 2]
```diff
@@ -1,37 +1,145 @@
...
+            Verify MaaS,
+            Install Tempest,
+            Tempest Tests,
+            Prepare Kibana Selenium,
+            Kibana Tests,
+            Holland
+      - leapfrogupgrade:
+          ACTION_STAGES: >-
+            Prepare MaaS,
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: same here

File: rpc_jobs/multi_node_aio.yml [hunk 3]
```diff
@@ -1,37 +1,145 @@
...
+      # as the upgrade method is not supported for any
+      # other target series.
+      # the kilo --> newton141 job is triggered by PRs to either branch.
+      - series: liberty
+        action: leapfrogupgrade
+      - series: mitaka
+        action: leapfrogupgrade
+      - series: newton140
+        action: leapfrogupgrade
+      - series: master
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: master + leapfrog is, I think, now redundant with xenial + leapfrog upgrade.

File: rpc_jobs/multi_node_aio.yml [hunk 4]
```diff
@@ -1,37 +1,145 @@
...
+        action: leapfrogupgrade
+      # Leapfrog upgrades cannot be executed on
+      # xenial as it is not possible to install
+      # the source series (kilo-mitaka) on xenial.
+      - image: xenial
+        action: leapfrogupgrade
+      # Xenial builds are run for newton and above
+      # as it is not supported distro before newton.
+      - series: mitaka
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: that looks like a duplicate.

File: rpc_jobs/multi_node_aio.yml [hunk 5]
```diff
@@ -49,9 +157,9 @@
         stage('OnMetal Multi-Node AIO') {{
           git branch: "{branch}", url: "https://github.com/rcbops/rpc-openstack"
           if("{series}" == "mitaka"){{
-            build job: 'OnMetal-Multi-Node-AIO_mitaka-trusty-post-merge'
+            build job: 'OnMetal-Multi-Node-AIO_mitaka-trusty-deploy-post-merge'
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: Is only mitaka enough here?
Not so sure to understand the prupose of all of this.

File: rpc_jobs/multi_node_aio.yml [hunk 6]
```diff
@@ -66,8 +174,21 @@
...
+      Connect Slave,
+      Prepare Multi-Node AIO,
+      Prepare RPC Configs,
+      Deploy RPC w/ Script,
+      Cleanup,
+      Destroy Slave
+    ACTION_STAGES: ""
+    TRIGGER_STAGES: ""
+    DEPLOY_TELEGRAF: "NO"
+    DEPLOY_INFLUX: "NO"
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: see my remark above with DEPLOY_MAAS.

File: rpc_jobs/multi_node_aio.yml [hunk 7]
```diff
@@ -152,6 +291,17 @@
...
+                withEnv(common.get_deploy_script_env()) {{
+                  sh """#!/bin/bash -xe
+                     ./scripts/bootstrap-ansible.sh
+                     """
+                }}
+              }}
+            }}
             deploy.deploy_sh(
               environment_vars: [
                 "DEPLOY_HAPROXY=yes",
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: See remark with deploy_maas for L311 on the right.

File: rpc_jobs/multi_node_aio.yml (hunk 5 above)
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: I think these jobs are outside the scope of this work. I only changed what was required to allow them to continue to work. However, looking at them again I forgot to add the resource bit to the name so will update that.

File: rpc_jobs/multi_node_aio.yml (hunk 3 above)
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: I've copied this from rpc_aio.yml. While master might only be xenial now, I'd rather keep to to files consistent. It also doesn't hurt that it's explicit.

Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: agreed.

File: rpc_jobs/multi_node_aio.yml (hunk 4 above)
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: Fixed

**Comment by git-harry** [comment]
Successful job: https://rpc.jenkins.cit.rackspace.net/job/gh-OnMetal-Multi-Node-AIO_newton141-trusty-leapfrogupgrade-small-periodic/38/

**Comment by mancdaz** [comment]
The code has 2 good reviews, and any failures in tests are not to do with the code in this PR. Merging, and we can create new tickets to fix any new problems we might encounter.
